heareval/predictions/task_predictions.py

```diff
@@ -79,6 +82,117 @@ def __getitem__(self, idx):
         # return self.embedding_memmap[idx], self.labels[idx]
 
 
+def create_events_from_prediction(
+    prediction_dict: Dict, threshold: float = 0.5
+) -> IntervalTree:
+    """
+    Takes a set of prediction tensors keyed on timestamps and generates events.
+
+    Args:
+        predictions: A dictionary of predictions {timestamp -> prediction}
+        threshold: Threshold for determining whether to apply a label
+
+    Returns:
+        An IntervalTree object containing all the events from the predictions.
+    """
+    # Make sure the timestamps are in the correct order
+    timestamps = sorted(prediction_dict.keys())
+
+    # Create a sorted tensor of frame level predictions for this file.
+    predictions = torch.stack([prediction_dict[t] for t in timestamps])
+
+    # Apply thresholding to get the label detected at each frame.
+    # TODO: Apply median filtering for smoothing?
+    #   Additionally, DCASE limits the number of active classes at a single
+    #   time to 6. We could do that here as well.
+    predictions = (predictions > threshold).type(torch.int8)
+
+    # Difference between each timestamp to find event boundaries
+    pred_diff = torch.diff(predictions, dim=0)
+
+    # This is slow, but for each class look through all the timestamp predictions
+    # and construct a set of events with onset and offset timestamps.
```

Comment:
I really am not sure what this is doing, is it what `more_itertools.consecutive_groups` does? https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.consecutive_groups

Reply:
It's based on the dcase code - it's basically looking for boundaries between consecutive groups of the *same* element. It achieves that by using the difference function above. So you get a value of 1 when you have an onset and then -1 for an onset because we have converted to binary vectors. So it starts at the first timestamp and loops through looking for the boundaries and adds an event when there has been an onset and then an offset.